### PR TITLE
[ALLUXIO-3159] Sync plan change so that only content change causes delete and reload

### PR DESCRIPTION
**Pull request by yuzhu**
https://alluxio.atlassian.net/browse/ALLUXIO-3159

If non-content related attributes have changed in the UFS file, the corresponding Alluxio file should just be updated with that info, and not be deleted if not necessary.

### CODE REVIEW

File: core/common/src/main/java/alluxio/underfs/Fingerprint.java [hunk 1]
```diff
@@ -118,6 +118,15 @@ public static Fingerprint parse(String input) {
     return new Fingerprint(tagMap);
   }
 
+  /**
+   * Checking if the fingerprint object was generated from an INVALID_UFS_FINGERPRINT.
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/Fingerprint.java]: NIT: "Checks"

this is the convention for method comments.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 2]
```diff
@@ -110,13 +130,15 @@ public static boolean inodeUfsIsSynced(Inode inode, String ufsFingerprint) {
    */
   public static final class SyncPlan {
     private boolean mUpdateDirectory;
+    private boolean mUpdateMetadata;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: can we just have `mUpdateMetadata` and get rid of `mUpdateDirectory`? Is `mUpdateDirectory` just the directory version of this feature?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 3]
```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
...
+   * @return true of the inode is synced with the ufs status
+   */
+  public static boolean inodeUfsIsMetadataSynced(Inode inode, Fingerprint ufsFingerprint) {
+    Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
+    return cachedFingerprint.matchMetadata(ufsFingerprint);
+  }
+
   /**
    * A class describing how to sync an inode with the ufs.
    * A sync plan has several steps:
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: Can you update these steps with the new version of the steps?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 4]
```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
...
 
+  /**
+   * Returns true if the given inode's metadata matches the ufs fingerprint.
+   *
+   * @param inode the inode to check for sync
+   * @param ufsFingerprint the ufs fingerprint to check for the sync
+   * @return true of the inode is synced with the ufs status
+   */
+  public static boolean inodeUfsIsMetadataSynced(Inode inode, Fingerprint ufsFingerprint) {
+    Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: multiple spaces after `=`.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 5]
```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
...
     boolean isSyncedUnpersisted =
-        !inode.isPersisted() && Constants.INVALID_UFS_FINGERPRINT.equals(ufsFingerprint);
-
+        !inode.isPersisted() && !ufsFingerprint.isValid();
     boolean isSyncedPersisted;
+
     if (inode instanceof InodeFile) {
       // check the file fingerprint.
       InodeFile inodeFile = (InodeFile) inode;
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inodeFile.getUfsFingerprint());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: NIT: I think I prefer `inodeFingerprint`?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 6]
```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
...
       isSyncedPersisted = inodeFile.isPersisted()
-          && inodeFile.getUfsFingerprint().equals(ufsFingerprint)
-          && !inodeFile.getUfsFingerprint().equals(Constants.INVALID_UFS_FINGERPRINT);
+          && cachedFingerprint.matchContent(ufsFingerprint)
+          && cachedFingerprint.isValid();
     } else {
-      isSyncedPersisted = inode.isPersisted() && inode.getUfsFingerprint().equals(ufsFingerprint);
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
+      isSyncedPersisted = inode.isPersisted() && cachedFingerprint.matchContent(ufsFingerprint);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: Do you think we can combine the dir and file paths to be the same, now that we have better fingerprint functionality?

File: tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java [hunk 7]
```diff
@@ -482,6 +482,41 @@ public void ufsDirUpdatePermissions() throws Exception {
...
+    status = mFileSystem.getStatus(new AlluxioURI(alluxioPath(EXISTING_FILE)), options);
+    Assert.assertNotNull(status);
+
+    // Make sure the mode is correctly updated with a metadata change only
+    Assert.assertEquals(
+        FileUtils.translatePosixPermissionToMode(PosixFilePermissions.fromString("rwxrwxrwx")),
+        status.getMode());
+    long afterFileid = status.getFileId();
+    // Change the permission of the file and the file id should not change
+    Assert.assertEquals(prevFileid, afterFileid);
```
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java]: NIT: `afterFileid` can just be `status.getFileId()`.

File: tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java [hunk 8]
```diff
@@ -482,6 +482,41 @@ public void ufsDirUpdatePermissions() throws Exception {
...
+
+    // Change the content of the file and the file id should change as a result because it is
+    // deleted and reloaded.
+    writeUfsFile(ufsPath(EXISTING_FILE), 2);
+
+    status = mFileSystem.getStatus(new AlluxioURI(alluxioPath(EXISTING_FILE)), options);
+    Assert.assertNotNull(status);
+    long afterContentChangeFileid = status.getFileId();
+    Assert.assertNotEquals(prevFileid, afterContentChangeFileid);
```
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java]: NIT: NIT: `afterContentChangeFileid ` can just be `status.getFileId()`.

File: core/common/src/main/java/alluxio/underfs/Fingerprint.java (hunk 1 above)
Comment (yuzhu) [reply core/common/src/main/java/alluxio/underfs/Fingerprint.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 2 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 3 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 4 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 5 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 6 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done

File: tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java (hunk 7 above)
Comment (yuzhu) [reply tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java]: done

File: tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java (hunk 8 above)
Comment (yuzhu) [reply tests/src/test/java/alluxio/master/file/UfsSyncIntegrationTest.java]: done

**Comment by yuzhu** [comment]
@gpang  comments addressed, PTAL.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 9]
```diff
@@ -34,95 +33,110 @@ private UfsSyncUtils() {} // prevent instantiation
...
+  /**
+   * Returns true if the given inode's metadata matches the ufs fingerprint.
+   *
+   * @param inode the inode to check for sync
+   * @param ufsFingerprint the ufs fingerprint to check for the sync
+   * @return true of the inode is synced with the ufs status
+   */
+  public static boolean inodeUfsIsMetadataSynced(Inode inode, Fingerprint ufsFingerprint) {
+    Fingerprint inodeFingerprint = Fingerprint.parse(inode.getUfsFingerprint());
+    return inodeFingerprint.matchMetadata(ufsFingerprint);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: do we have to check for `isValid`?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java [hunk 10]
```diff
@@ -34,95 +33,110 @@ private UfsSyncUtils() {} // prevent instantiation
...
+        // Only update the inode if it is not the root directory. The root directory is a special
+        // case, since it is expected to be owned by the process that starts the master, and not
+        // the owner on UFS.
+        syncPlan.setUpdateMetadata();
+      }
+      syncPlan.setSyncChildren();
+      return syncPlan;
+    }
+
+    // One of metadata or content is not in sync and it is a file
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: Can a directory reach this point too?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 9 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: done.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java (hunk 10 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java]: The only way for a directory to reach this point, is when the inode says it is a directory and ufs says it is a file .  In which case, we do the delete and reload anyways.  I modified the comment to reflect that.

**Comment by yuzhu** [comment]
@gpang  Thanks for the review.  Addressed comments, PTAL.

**Pull request merged**